Add server streaming tests.

marshall-lee opened the pull request:
Hello!

I plan to use grpc-haskell in my project so I want to help.
Here I try to add some missing server streaming tests.

@kolmodin, please respond if you're interested in my help.

kolmodin commented on tests/interop_client/main.hs:
```diff
@@ -506,6 +516,83 @@ runClientStreamingTest opts = do
+        RpcError err ->
+          return (Left (show err))
+
+runServerStreamingWithSlowConsumerTest :: Options -> IO (Either String ())
```
This test is not part of the interop testing suite.
It's also far too slow to have as a test here, it would make `--test_case=all` take very long.

But do feel free to move this to it's own test file, I propose
`tests/slow_consumer/main.hs`. Thanks!

kolmodin replied:
Oh, I see, it's part of the C++ client for interop tests.
Then please also do what they do there, please change so that it doesn't run when `--test_case=all`.

`testCaseMap` can have type `[(String, Bool, TestCase)]` where the bool is `True` for tests that should run with `--test_case=all`.
Let `allTests` return the tests that have `True`.

kolmodin commented on tests/interop_client/main.hs:
```diff
@@ -506,6 +516,83 @@ runClientStreamingTest opts = do
+    responsesCount = 2000
+    delaySeconds = 20
+    delay = delaySeconds * 1000
+    req = def { _StreamingOutputCallRequest'responseParameters = map (\_ -> def { _ResponseParameters'size = responseSize }) [1 .. responsesCount] }
```
Please split this into two lines.

kolmodin replied:
Please use `replicate`.

kolmodin commented on tests/interop_client/main.hs:
```diff
@@ -506,6 +516,83 @@ runClientStreamingTest opts = do
+        closeCall
+        return msgs
+      case resp of
+        RpcOk result ->
```
This would be better written as;

```haskell
        RpcOk (Right msgs)
          | length msgs == responsesCount -> return (Right ())
          | otherwise -> return (Left ("responses count does not match: " ++ (show (length msgs))))
        RpcOk (Left err) ->
          return (Left err)
```

marshall-lee commented:
@kolmodin 

Hi again! I hope I fixed all the issues you mentioned. I used guards where it fits good, put some computations inside `Either` monad and removed some long lines.

kolmodin commented:
@marshall-lee  
Awesome, thanks! I'll merge it. And congrats to being the first contributor :)

С Новым годом!

marshall-lee commented:
@kolmodin lol I'm the first contributor of grpc-haskell and this is my first sensible haskell pull request in general.
Happy new year!
